### PR TITLE
Implement benchmarking using Criterion

### PR DESCRIPTION
**Pull request by mishig25**
### TLDR: add benchmarking for `serialization` & `deserialziation`

Not sure if we will do parallelization https://github.com/huggingface/safetensors/issues/15
But either way, I thought it is probably a good idea to add some benchmarking.

I'm using [criterion](https://github.com/bheisler/criterion.rs), which is a popular rust benchmarking framework (also used on [tokenizers](https://github.com/huggingface/tokenizers/blob/0f20b0823a41de7aa918e90e58a3398b024f4b73/tokenizers/Cargo.toml#L76)).

Can be run as:
```
cargo bench
```
Produces output as:
```
Serlialize 10_MB
time:   [427.15 µs 436.21 µs 446.56 µs]                             
change: [-0.3956% +0.7567% +1.9834%] (p = 0.32 > 0.05)
No change in performance detected.

Deserlialize 10_MB                     
time:   [5.5303 µs 5.5634 µs 5.5997 µs]                                
change: [-0.4513% +0.5147% +1.5265%] (p = 0.32 > 0.05)
No change in performance detected.
```
Read more about [cargo bench here](https://doc.rust-lang.org/cargo/commands/cargo-bench.html).
Note: since criterion is a dev-dependency, it does not get added in the prod build

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
For what my Rust beginner's opinion is worth, LGTM!
Thanks for adding benchmarks!

**Review by Narsil — APPROVED** [review]
LGTM.

Very nice !
Just FYI, there are some "speed" tests directly in Python.
For this sort of job, the disk access is likely to be an extremely important part of the overall speed of the program so it's important to keep that in mind.

Btw, if you want to check an idea for speeding things up, I think it could be in the load order of tensors. (Only PT is really faster afaik).

If we could order `safe_open("..").keys()` to be in increasing offsets, it's likely that it's going to be faster to load the overall array.
I haven't really tested in, but the speeds are so flaky right now that I'm guessing part of it is the randomization linked to the hashmap.

**Comment by mishig25** [comment]
> here are some "speed" tests directly in Python.
@Narsil could you elaborate on this? 🤗 

> For this sort of job, the disk access is likely to be an extremely important part of the overall speed of the program so it's important to keep that in mind.

Is there anything I should do in this PR or is it good to merge?

**Comment by Narsil** [comment]
`pytest -sv tests/`  within `bindings/python` and you can see some performance measurements. (Fully serializing a gpt2 model to disk and back).

**Comment by Narsil** [comment]
> Is there anything I should do in this PR or is it good to merge?

Go right ahead ! Thanks !

**Comment by mishig25** [comment]
> pytest -sv tests/ within bindings/python and you can see some performance measurements. (Fully serializing a gpt2 model to disk and back).

oh I see, thanks a lot! I was looking specifically for a test with name `*speed*`
